lib/build.ts

```diff
@@ -0,0 +1,99 @@
+import * as Promise from 'bluebird';
+import * as Dockerode from 'dockerode';
+import * as _ from 'lodash';
+import { Builder, BuildHooks } from 'resin-docker-build';
+import * as Stream from 'stream';
+
+import { BuildTask, Dict } from './build-task';
+import { BuildProcessError } from './errors';
+import { pullExternal } from './external';
+import { LocalImage } from './local-image';
+
+function taskHooks(
+	task: BuildTask,
+	docker: Dockerode,
+	resolve: (image: LocalImage) => void,
+): BuildHooks {
+	return {
+		buildSuccess: (imageId: string, layers: string[]) => {
+			Promise.try(() => {
+				if (task.tag != null) {
+					return Promise.resolve(docker.getImage(imageId).tag({ repo: task.tag, force: true }))
+						.return(task.tag);
+				} else {
+					return imageId;
+				}
+			})
+			.then((tag) => {
+				const image = new LocalImage(docker, tag, task.serviceName, false, true);
+				image.layers = layers;
+
+				resolve(image);
+			});
+		},
+		buildFailure: (error: Error, layers: string[]) => {
+			const image = new LocalImage(
+				docker,
+				layers[layers.length  - 1],
+				task.serviceName,
+				false,
+				false,
```

Comment:
Booleans are slightly better when passed as an object because it makes clear what they're for.